### PR TITLE
Added beamsearch_set_finished Operator

### PR DESCRIPTION
**Pull request by blchu**
## Description ##
Added a new operator, beamsearch_set_finished, for updating the scores of finished beams in beam search more efficiently than using alternate operators.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] New operator, beamsearch_set_finished, and appropriate unit test

## Comments ##
- Operator will be added to contrib

### CODE REVIEW

File: src/operator/contrib/beamsearch_set_finished-inl.h [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    CHECK_EQ(outputs.size(), 1U);
+    CHECK_EQ(req.size(), 1U);
+
+    const mxnet::TShape& out_shape = outputs[beamsearch_set_finished::kOut].shape_;
+    const mxnet::TShape& batch_beam_shape = inputs[beamsearch_set_finished::kFin].shape_;
+
+    mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+    using namespace mxnet_op;
+    MSHADOW_TYPE_SWITCH(outputs[beamsearch_set_finished::kOut].type_flag_, DType, {
+        MSHADOW_TYPE_SWITCH(inputs[beamsearch_set_finished::kFin].type_flag_, IType, {
```
Comment (haojin2) [inline src/operator/contrib/beamsearch_set_finished-inl.h]: seems like you are limiting the data type of `Fin` and `OverMax` to `int32` only in the `InferType` function, you then do not need a type switch here for those guys, you can simply do:
```c++
MSHADOW_TYPE_SWITCH(outputs[beamsearch_set_finished::kOut].type_flag_, DType, {
  DType mask_val = param.mask_val;
  const int score_idx = param.score_idx;
  const int eos_idx = param.eos_idx;
  Kernel<beamsearch_set_finished_forward, xpu>::Launch(
    s, out_shape.Size(),
    outputs[beamsearch_set_finished::kOut].dptr<DType>(),
    inputs[beamsearch_set_finished::kDist].dptr<DType>(),
    inputs[beamsearch_set_finished::kScores].dptr<DType>(),
    inputs[beamsearch_set_finished::kFin].dptr<int32>(),
    inputs[beamsearch_set_finished::kOverMax].dptr<int32>(),
    mask_val, score_idx, eos_idx, out_shape.Size()/batch_beam_shape.Size());
});
```

File: src/operator/contrib/beamsearch_set_finished.cc [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+.add_argument("finished", "NDArray-or-Symbol", "Finished beams")
+.add_argument("over_max", "NDArray-or-Symbol", "Beams at or exceeding maximum length")
+.add_arguments(BeamsearchSetFinishedParam::__FIELDS__());
+
+NNVM_REGISTER_OP(_contrib_beamsearch_noop_grad)
+.set_num_inputs(1)
+.set_num_outputs(4)
+.set_attr<FCompute>("FCompute<cpu>", NoopGrad<cpu>)
+.set_attr<nnvm::TIsBackward>("TIsBackward", true);
+}
```
Comment (haojin2) [inline src/operator/contrib/beamsearch_set_finished.cc]: one more blank line below.

File: tests/python/gpu/test_operator_gpu.py [hunk 3]
```diff
@@ -1698,6 +1698,54 @@ def test_take_with_type():
                                         'take_a': 'write'},
                               arg_params=arg_params)
 
+@with_seed()
```
Comment (haojin2) [inline tests/python/gpu/test_operator_gpu.py]: one more blank line above

File: src/operator/contrib/beamsearch_set_finished-inl.h (hunk 1 above)
Comment (blchu) [reply src/operator/contrib/beamsearch_set_finished-inl.h]: I see, I'll change that then

**Comment by sxjscience** [comment]
Is it possible to automatically fuse these operators via the recently introduced point-wise fusion?

**Comment by blchu** [comment]
Currently there isn't support for fusing concatenation and broadcasting operators, though I'm not sure about the feasibility of doing so in the future.

**Pull request closed without merging**